src/main/java/com/spotify/confidence/EventSenderEngineImpl.java

```diff
@@ -44,13 +52,34 @@ public void run() {
         }
         final var numberOfPendingEvents = eventStorage.write(event);
         if (numberOfPendingEvents >= maxBatchSize) {
-          eventStorage.createBatch();
-          uploadQueue.add(UPLOAD_SIG);
+          flush();
+        } else {
+          scheduleFlush();
         }
       }
     }
   }
 
+  private void scheduleFlush() {
+    // Cancel the existing scheduled task if it exists
+    if (pendingFlush != null && !pendingFlush.isDone()) {
+      pendingFlush.cancel(false);
+    }
+    if (flushTimeoutMilliseconds > 0) {
+      pendingFlush =
+          executorService.schedule(this::flush, flushTimeoutMilliseconds, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void flush() {
```

Comment:
What happens if this is triggered both from `numberOfPendingEvents >= maxBatchSize` and by the `flushTimeoutMilliseconds` at the same time?

Reply:
Perhaps you can replace the `BlockingQueue<String> uploadQueue` with a `BlockingQueue<EventBatch> uploadQueue`, and then the UploadPoller thread can just take batches from that? 

So flush() would do something like: 
```
uploadQueue.add(eventStorage.createBatch());
```

in both the timer triggered case and the size triggered case.

Reply:
Like we noticed in call @nicklasl createBatch is synchronous, which means there is no concurrency issue. It does however mean we might briefly block the callers thread (sending event).. I think we can live with that for now and create a followup PR where we revisit the queuing strategy, and maybe do what @DennisPersson is suggesting.